### PR TITLE
Allow mach build to explicitly set the media stack

### PR DESCRIPTION
**Pull request by asajeffrey**
This allows `./mach build` to override the media stack, which is currently hard-wired based on the target.

---

- [X] `./mach build -d` does not report any errors
- [X]  `./mach test-tidy` does not report any errors
- [X] These changes do not require tests because it's build infra

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23423)

### CODE REVIEW

**Comment by asajeffrey** [comment]
The step back is that running raw cargo commands like `cargo check` will now fail because there's no `media-foo` feature. The alternative is to use default features, but that has its own pain points, since you can only switch off all default features, there's no way to switch off one default.

File: ports/glutin/Cargo.toml [hunk 1]
```diff
@@ -34,6 +34,8 @@ max_log_level = ["log/release_max_level_info"]
 webdriver = ["libservo/webdriver"]
 energy-profiling = ["libservo/energy-profiling"]
 debugmozjs = ["libservo/debugmozjs"]
+media-dummy = ["libservo/servo-media-dummy"]
+media-gstreamer = ["libservo/servo-media-gstreamer"]
```
Comment (jdm) [inline ports/glutin/Cargo.toml]: I suspect we need these in libsimpleservo as well.

**Comment by asajeffrey** [comment]
Another option would be to build `servo-media-dummy` for all platforms, and just only link it on platforms where gstreamer isn't available.

**Review by jdm — COMMENTED** [review]
Another option that would keep the platform logic inside Cargo might be a dummy-media-backend feature, and the cfgs could be `any(feature = dummy-media-backend, (os stuff))` and `all(not(feature = dummy-media-backend), (os stuff))`.

**Comment by asajeffrey** [comment]
The motivation for this change is so that we have a brute force solution to people who are hitting problems like #21970, which are quite frustrating if you don't need gstreamer.

**Comment by afcady** [comment]
Note that another solution, or at least workaround, for #21970 is just to specify `RUSTFLAGS='-L ./support/linux/gstreamer/gst/lib'`.

**Comment by asajeffrey** [comment]
Yeah, that works for that particular issue, but we keep hitting build problems with gstreamer, it might be nice to have a workaround.

**Comment by ferjm** [comment]
Thanks @asajeffrey! We definitely need something like this, at least until we have an easier way to setup gstreamer.

I like @jdm's [suggestion](https://github.com/servo/servo/pull/23423#pullrequestreview-239110643) but I suspect that we'd still have the `cargo check` issue that you [mentioned](https://github.com/servo/servo/pull/23423#issuecomment-493581105), right?

**Comment by tuncer** [comment]
@asajeffrey, I have a similar use case and would like to see this in master. Would it help if I attempted a fresh rebase?

**Comment by asajeffrey** [comment]
@tuncer sure!

**Comment by tuncer** [comment]
@asajeffrey, ok, I'll let you know if I run into roadblocks.

**Comment by tuncer** [comment]
@asajeffrey, sorry for the delay: [new diff](https://github.com/servo/servo/compare/master...tuncer:configure-media-stack-v2) ([patch file](https://github.com/servo/servo/compare/master...tuncer:configure-media-stack-v2.patch)). Haven't tried to build yet.

**Comment by asajeffrey** [comment]
Rather annoyingly I don't think github allows me to make you the owner of this PR  :( (https://stackoverflow.com/questions/50781553/how-to-take-ownership-of-a-pull-request has the answer "pretty much nope")

Can you open a new PR with your changes and reference this one? Also cc me in a comment in the new PR (but not the initial commit as that is a recipe for chaos!)

**Comment by tuncer** [comment]
I don't mind doing that, but have you considered replacing your local branch with the rebased one and running `git push origin +configure-media-stack`?

**Comment by asajeffrey** [comment]
I could, but it means I have to do that for every edit you make, which I think both of us would find kind of frustrating after a while.

**Comment by tuncer** [comment]
I didn't intend to take ownership of the patch but just resolve the merge conflict, as you said you'd welcome that. Hence, I made sure to retain git author and commit date.

**Comment by tuncer** [comment]
@asajeffrey, done, see #24668.

**Comment by asajeffrey** [comment]
Replaced by #24668

**Pull request closed without merging**